[Converters] Adding updated samples

niels9001 opened the pull request:
- Adding a few new samples for various converters. (No sample for e.g. TaskResultConverter and ResourceStringConverter as these might not be used that much?) Maybe a code 1 liner is enough?

Known issues:
- The EmptyCollection converter sample does not seem to work.. Converter doesn't get triggered when the collection changes?
- ColorToDisplayName sample not working for WASDK

@michael-hawker Would love to get your feedback on these samples - wanted to keep them as simple as possible to let devs quickly copy-and-paste what they need vs. complex use cases - especially since Converters are pretty straightforward to work with?

michael-hawker commented:
Weird different test failure:
```
 Failed Test_RichSuggestBox_AddTokens (@Token1,@Token2,@Token3) [142 ms]
  Error Message:
   Assert.IsTrue failed. SuggestionChosen was not invoked.
  Stack Trace:
     at RichSuggestBoxExperiment.Tests.Test_RichSuggestBox.TestAddTokenAsync(RichSuggestBox rsb, String tokenText) in D:\a\Windows\Windows\components\RichSuggestBox\tests\Test_RichSuggestBox.cs:line 354
   at RichSuggestBoxExperiment.Tests.Test_RichSuggestBox.<>c__DisplayClass0_0.<<Test_RichSuggestBox_AddTokens>b__0>d.MoveNext() in D:\a\Windows\Windows\components\RichSuggestBox\tests\Test_RichSuggestBox.cs:line 33
--- End of stack trace from previous location ---
   at CommunityToolkit.WinUI.DispatcherQueueExtensions.<>c__DisplayClass3_0.<<EnqueueAsync>b__1>d.MoveNext()
--- End of stack trace from previous location ---
   at RichSuggestBoxExperiment.Tests.Test_RichSuggestBox.Test_RichSuggestBox_AddTokens(String tokenText1, String tokenText2, String tokenText3) in D:\a\Windows\Windows\components\RichSuggestBox\tests\Test_RichSuggestBox.cs:line 26
   at Microsoft.VisualStudio.TestPlatform.MSTestAdapter.PlatformServices.ThreadOperations.ExecuteWithAbortSafety(Action action)
```
Reran job

michael-hawker commented:
@niels9001 realizing that the color converter directly throwing an exception on WASDK is going to mean we can't load any of the samples on this page in our sample app as it just takes down the app. That's not really good... 😋

ChaseKnowlden commented:
I'm watching @michael-hawker's stream.

michael-hawker commented:
@niels9001 for the `EmptyCollectionToObjectConverter` base class it's not dynamic, it only cares about the initial state. Otherwise it'd need to listen to the `INotifyCollectionChanged` interface on the collection, but that sort of connection isn't something a converter is meant to do. It's really meant for a page that initializes a collection based on some result and not for a dynamic collection.

We should update the comments on the classes and the documentation here as such.

For the dynamic scenario the `IsNullOrEmptyStateTrigger` is probably going to be the way to go? So, we should call that out here? 

Though another solution may be able to be an x:Bind function that takes the `Items.Count`? As I think that should respond to changes...

michael-hawker commented on components/Converters/samples/BoolNegationConverterSample.xaml:
```diff
@@ -0,0 +1,17 @@
+        <converters:BoolNegationConverter x:Key="BoolNegationConverter" />
+    </Page.Resources>
+
+    <ToggleSwitch Header="Converted value"
```
Should this just be textbox or disabled or something else? It's weird you can interact with the ToggleSwitch here when the example is to toggle the one in the option pane and see it negated, right?

michael-hawker commented on components/Converters/samples/BoolToVisibilityConverterSample.xaml:
```diff
@@ -0,0 +1,27 @@
+      mc:Ignorable="d">
+
+    <Page.Resources>
+        <converters:BoolToVisibilityConverter x:Key="BoolToVisibilityConverter"
+                                              FalseValue="Collapsed"
+                                              TrueValue="Visible" />
```
Are these not the default values?

Would probably be better to show the parameter being used to invert the effect (like have two images, one that's shown and one that's hidden) that way it shows the default usage as well as the inverted pattern with the `ConverterParameter`?

michael-hawker commented on components/Converters/samples/ColorToDisplayNameConverterSample.xaml:
```diff
@@ -0,0 +1,32 @@
+                Spacing="16">
+
+        <ColorPicker x:Name="colorPicker"
+                     Height="64"
```
I guess this is meant to be cut-off? It looks a little odd with the shape:

![image](https://github.com/CommunityToolkit/Windows/assets/24302614/a1a8a41e-332e-441c-96fa-1335dc65d2a1)

Thoughts on if we return a "Not Supported" string instead of throwing an exception?

michael-hawker commented on components/Converters/samples/DoubleToVisibilityConverterSample.xaml:
```diff
@@ -0,0 +1,28 @@
+      mc:Ignorable="d">
+
+    <Page.Resources>
+        <converters:DoubleToObjectConverter x:Key="DoubleToVisibilityConverter"
+                                            FalseValue="Collapsed"
+                                            GreaterThan="5"
+                                            TrueValue="Visible" />
```
Again, let's not set the default values within the XAML when they're the defaults.

Since this is similar to the DoubleToObject example (which I love showing the slider with both greater and less), maybe we use the ConverterParameter to invert in this one?

niels9001 replied:
Aaaah.. just realized I was using a `DoubleToObjectConverter` vs. a `DoubleToVisibilityConverter`, hence why it crashes when the `FalseValue` or `TrueValue` wasn't set.

Oops - copy+paste error :(. Fixed now!

michael-hawker commented on components/Converters/samples/CollectionVisibilityConverterSample.xaml:
```diff
@@ -0,0 +1,40 @@
+            <RowDefinition Height="Auto" />
+        </Grid.RowDefinitions>
+
+        <ListView Header="Lalala"
+                  ItemsSource="{x:Bind Items, Mode=OneWay}"
+                  Visibility="{x:Bind Items, Mode=OneWay, Converter={StaticResource CollectionVisibilityConverter}}" />
+
+        <TextBlock Text="No items in this list"
+                   Visibility="{x:Bind Items, Mode=OneWay, Converter={StaticResource CollectionVisibilityConverter}}" />
```
One of these needs to be inverted, eh?

michael-hawker commented on components/Converters/samples/CollectionVisibilityConverterSample.xaml.cs:
```diff
@@ -0,0 +1,32 @@
+
+    private void Add_Click(object sender, RoutedEventArgs e)
+    {
+        Items.Clear();
+        Items.Add("Item 1");
+        Items.Add("Item 2");
+        Items.Add("Item 3");
+    }
+
+    private void Clear_Click(object sender, RoutedEventArgs e)
+    {
+        Items.Clear();
```
To make this 'work' in the sample, instead of clearing and adding items, just replace `Items` with a new collection. Again, we'll want to call this out in the documentation like I added in the other comment on the PR, but this should get it to behave 'as expected'.

niels9001 replied:
Aaah.. well, might make more sense to show e.g. a placeholder text on an initial app load? Adopted the Outlook message for now:

![image](https://github.com/CommunityToolkit/Windows/assets/9866362/a6508395-40fd-476c-941a-9965716c9b2f)

niels9001 replied:
Added a note on using the trigger.. let me know if that suffices.

michael-hawker commented on components/Converters/samples/CollectionVisibilityConverterSample.xaml.cs:
```diff
@@ -0,0 +1,32 @@
+    {
+        this.InitializeComponent();
+        Items = new();
+
+
```
Extra whitespace

michael-hawker commented on components/Converters/samples/EmptyStringToObjectConverterSample.xaml:
```diff
@@ -0,0 +1,23 @@
+    </Page.Resources>
+
+    <StackPanel>
+        <Button Content="Submit"
```
Could be nice to have the button do something? I know it's silly, but it is unsatisfying... 😋

michael-hawker commented on components/Converters/samples/Converters.md:
```diff
@@ -12,151 +12,71 @@ issue-id: 0
+> [!Sample EmptyStringToObjectConverterSample]
 
-## BoolToVisibilityConverter Examples
+## FileSizeToFriendlyStringConverter
```
This one is missing a sample, maybe could simulates a progress bar 'download' with increasing size to a target value that goes when you click a button... That may be a bit complicated, but would make an interesting/realistic sample...

niels9001 replied:
Yeah, was thinking about this.. tried to add a slider option as a "filesize in bytes" thing.. but converting that to a `long` would require a `ConvertToLong` method in the XAML string, which might be confusing..

Added a simple sample that just parses a defined `long`:

![image](https://github.com/CommunityToolkit/Windows/assets/9866362/038e9bc1-21b5-4371-8dbe-88260cfeff93)

michael-hawker commented on components/Converters/samples/StringVisibilityConverterSample.xaml:
```diff
@@ -0,0 +1,21 @@
+      mc:Ignorable="d">
+
+    <Page.Resources>
+        <converters:StringVisibilityConverter x:Key="StringVisibilityConverter"
+                                              EmptyValue="Visible"
+                                              NotEmptyValue="Collapsed" />
```
Are these defaults or inverted? If inverted, should we use ConverterParameter instead?

Since many of these converters are similarly built/have base classes, maybe we need a section at the top with some of these general implementation notes. e.g. either can change values for true/false empty/notempty etc... or can use converterparameter to negate result...

michael-hawker commented on components/Converters/samples/Converters.md:
````diff
@@ -12,151 +12,71 @@ issue-id: 0
-<TextBlock Text="No Items." Visibility="{Binding Path=MyCollectionValue, Converter={StaticResource CollectionVisibilityConverter}, ConverterParameter=True}">
-```
+## TaskResultConverter
+Converter that can be used to safely retrieve results from `Task<TResult>` instances. See MVVM Toolkit Gallery for an example.
````
Link to the store download of the MVVM Toolkit Gallery?

michael-hawker commented on components/Converters/samples/TypeToObjectConverterSample.xaml:
```diff
@@ -0,0 +1,22 @@
+                                          Type="x:String" />
+    </Page.Resources>
+
+    <TextBlock Foreground="{x:Bind MyTextStringValue, Mode=OneWay, Converter={StaticResource TypeToObjectConverter}}"
```
This one didn't work for me? @Arlodotexe thoughts on a better example for this one?

niels9001 replied:
Oh, forgot to add that to the known issues.. yeah, having a valid scenario would be helpful to improve this sample!

michael-hawker replied:
Think this is the main one left that's wonky. The original sample used the TextToolbar which kind of made sense. I could see maybe a shared generic DataTemplate type scenario being a case, but searching GitHub I didn't see any usages of this converter... hmmm

michael-hawker replied:
Added a working example, albeit still not practical, good enough for now.

michael-hawker commented on components/Converters/samples/Converters.md:
```diff
@@ -12,151 +12,71 @@ issue-id: 0
-
-> [!Sample StringFormatConverterSample]
+## VisibilityToBoolConverter
+Converts a `Visibility` enumaration to a boolean value.
```
Maybe for this one we just have a toggleswitch which toggles showing something, and then a textbox which binds to the visibility of that and outputs true/false to a TextBlock?

michael-hawker reviewed (commented):
Hey @niels9001 great job, love the micro samples. They're looking really good!

Have a few comments for improvements/nits and additions for a couple of missing ones. Also think I addressed your comments, and I have some fixes for some bugs I'm pushing to the tooling submodule, as I had issues testing/trying some things out.

michael-hawker commented:
Opened https://github.com/CommunityToolkit/Tooling-Windows-Submodule/pull/99 to fix the issue with the Converter reference which prevents modifying the converter code and testing it in the repo (as the Sample app instead always pulls from the NuGet package code instead of the local source).

niels9001 commented:
> Opened [CommunityToolkit/Tooling-Windows-Submodule#99](https://github.com/CommunityToolkit/Tooling-Windows-Submodule/pull/99) to fix the issue with the Converter reference which prevents modifying the converter code and testing it in the repo (as the Sample app instead always pulls from the NuGet package code instead of the local source).

Aah, good catch :-)! Yes, let's get that merged in so we can validate the change to the `ColorToDisplayNameConverter`.

I think I addressed most of the feedback - let me know what you think!

michael-hawker reviewed (approved):
Tested everything. Only things that don't seem to work well are the Color and Type converters on WASM, not sure why... I guess ColorPicker isn't supported on Uno? Not sure what's up with the types though there...

Other than that it's just waiting for WASDK to support the missing API for the Color Converter

michael-hawker commented:
Hmm, conflict with tooling submodule pointer, going to rebase locally...

michael-hawker commented:
Hmmm...
```
/home/runner/.nuget/packages/msbuild.sdk.extras/3.0.23/Build/Workarounds.targets(27,5): error : If you are building projects that require targets from full MSBuild or MSBuildFrameworkToolsPath, you need to use desktop msbuild ('msbuild.exe') instead of 'dotnet build' or 'dotnet msbuild' [/home/runner/work/Windows/Windows/components/Converters/src/CommunityToolkit.WinUI.Converters.csproj::TargetFramework=uap10.0.17763]
/home/runner/.nuget/packages/microsoft.windowsappsdk/1.3.230331000/buildTransitive/MrtCore.PriGen.targets(911,5): error MSB4062: The "Microsoft.Build.Packaging.Pri.Tasks.ExpandPriContent" task could not be loaded from the assembly /usr/share/dotnet/sdk/7.0.306/Microsoft/VisualStudio/v17.0/AppxPackage/Microsoft.Build.Packaging.Pri.Tasks.dll. Could not load file or assembly '/usr/share/dotnet/sdk/7.0.306/Microsoft/VisualStudio/v17.0/AppxPackage/Microsoft.Build.Packaging.Pri.Tasks.dll'. The system cannot find the file specified. [/home/runner/work/Windows/Windows/components/Converters/src/CommunityToolkit.WinUI.Converters.csproj::TargetFramework=net6.0-windows10.0.19041.0]
/home/runner/.nuget/packages/microsoft.windowsappsdk/1.3.230331000/buildTransitive/MrtCore.PriGen.targets(911,5): error MSB4062:  Confirm that the <UsingTask> declaration is correct, that the assembly and all its dependencies are available, and that the task contains a public class that implements Microsoft.Build.Framework.ITask. [/home/runner/work/Windows/Windows/components/Converters/src/CommunityToolkit.WinUI.Converters.csproj::TargetFramework=net6.0-windows10.0.19041.0]
    356 Warning(s)
    2 Error(s)
```
@Arlodotexe you see this before?

michael-hawker commented:
Actually probably has to do with the changes to the submodule and the Converters package... we do the same build in the Tooling repo though, but there's more components here... (including the source). Maybe missed something, will investigate locally.

michael-hawker commented:
Hmm, all three heads built fine locally for the generate all solution, so seems something specific to linux...

michael-hawker commented:
Rerunning the linux build, I tried locally on windows with the same build command and it was fine:

> dotnet build /r /bl /p:UnoSourceGeneratorUseGenerationHost=true /p:UnoSourceGeneratorUseGenerationController=false

```
    373 Warning(s)
    0 Error(s)

Time Elapsed 00:01:28.63

h:\code\CommunityToolkitWindows\tooling\ProjectHeads\AllComponents\Wasm>
```

Arlodotexe commented:
Needs https://github.com/CommunityToolkit/Tooling-Windows-Submodule/pull/112

michael-hawker commented:
This is waiting on #152 which was waiting on #160.

michael-hawker commented:
#152 is good now, so once that's merged we can get this cleaned-up/updated and in!

michael-hawker commented:
Rebased on `main`, should hopefully be good to go once it builds...
